orchagent/routeorch.cpp

```diff
@@ -2106,3 +2112,743 @@ bool RouteOrch::removeOverlayNextHops(sai_object_id_t vrf_id, const NextHopGroup
+        m_bulkNhgReducedRefCnt.clear();
+        while (it_prev != it)
+        {
+            KeyOpFieldsValuesTuple t = it_prev->second;
+
+            string key = kfvKey(t);
+            string op = kfvOp(t);
+            auto found = toBulk.find(make_pair(key, op));
+            if (found == toBulk.end())
+            {
+                it_prev++;
+                continue;
+            }
+
+            const auto& ctx = found->second;
+            const auto& object_statuses = ctx.object_statuses;
+            if (object_statuses.empty())
+            {
+                it_prev++;
+                continue;
+            }
+
+            const sai_object_id_t& vrf_id = ctx.vrf_id;
+            const Label& label = ctx.label;
+
+            if (op == SET_COMMAND)
+            {
+                const bool& excp_intfs_flag = ctx.excp_intfs_flag;
+
+                if (excp_intfs_flag)
+                {
+                    /* If any existing routes are updated to point to the
+                     * above interfaces, remove them from the ASIC. */
+                    if (removeLabelRoutePost(ctx))
+                        it_prev = consumer.m_toSync.erase(it_prev);
+                    else
+                        it_prev++;
+                    continue;
+                }
+
+                const NextHopGroupKey& nhg = ctx.nhg;
+
+                if (nhg.getSize() == 1 && nhg.hasIntfNextHop())
+                {
+                    if (addLabelRoutePost(ctx, nhg))
+                        it_prev = consumer.m_toSync.erase(it_prev);
+                    else
+                        it_prev++;
+                }
+                else if (m_syncdLabelRoutes.find(vrf_id) == m_syncdLabelRoutes.end() ||
+                         m_syncdLabelRoutes.at(vrf_id).find(label) == m_syncdLabelRoutes.at(vrf_id).end() ||
+                         m_syncdLabelRoutes.at(vrf_id).at(label) != nhg)
+                {
+                    if (addLabelRoutePost(ctx, nhg))
+                        it_prev = consumer.m_toSync.erase(it_prev);
+                    else
+                        it_prev++;
+                }
+            }
+            else if (op == DEL_COMMAND)
+            {
+                /* Cannot locate the route or remove succeed */
+                if (removeLabelRoutePost(ctx))
+                    it_prev = consumer.m_toSync.erase(it_prev);
+                else
+                    it_prev++;
+            }
+        }
+
+        /* Remove next hop group if the reference count decreases to zero */
+        for (auto& it_nhg : m_bulkNhgReducedRefCnt)
+        {
+            if (m_syncdNextHopGroups[it_nhg.first].ref_count == 0)
+            {
+                removeNextHopGroup(it_nhg.first);
+            }
+        }
+    }
+}
+
+void RouteOrch::addTempLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey &nextHops)
+{
+    SWSS_LOG_ENTER();
+
+    Label& label = ctx.label;
+
+    auto next_hop_set = nextHops.getNextHops();
```

Comment:
Is there any local cache for these temp label routes to process when all NHs are resolved?

Reply:
This code uses the same mechanism as existing Route code.